File: C#/1. Two Sum.cs
using System.Collections.Generic;
using System.Linq;
using System;

public class Solution {
    public int[] TwoSum(int[] nums, int target) {
    //Brute Force Approach
    //    int[] result = new int[2];
    //     for (int i = 0; i < nums.Length; i++) {
    //         for (int j = i + 1; j < nums.Length; j++) {
    //             if (nums[i] + nums[j] == target) {
    //                 result[0] = i;
    //                 result[1] = j;
    //                 return result;
    //             }
    //         }
    //     }
    //     return result;
    // }

    //Brute Force Approach
    // for (int i = 0; i < nums.Length; i++){
    //     for (int j = i + 1; j<nums.Length; j++){
    //         if ((nums[i] + nums[j]) == target){
    //              return new int[]{i,j};
    //         }
    //      }
    // }
    // return new int[]{};

    //Sorting and Two Pointers Approach
    //  (int num, int index)[] numbers = nums.Select((num, index) => (num, index)).ToArray();
    // Array.Sort(numbers, (a, b) => a.num.CompareTo(b.num));

    // int left = 0, right = nums.Length - 1;
    // while(left < right) {
    //     int sum = numbers[left].num + numbers[right].num;
    //     if(sum == target) {
    //         return new int[] { numbers[left].index, numbers[right].index };
    //     } else if(sum < target) {
    //         left++;
    //     } else {
    //         right--;
    //     }
    // }
    // return new int[] { };
  

    //Hash Table Appoach
    // Dictionary<int, int> map = new Dictionary<int, int>();
    // for (int i = 0; i < nums.Length; i++){
    //     int complement = target - nums[i];
    //     if (map.ContainsKey(complement)){
    //         return new int[]{map[complement], i};
    //     }
    //     if (!map.ContainsKey(nums[i])){
    //         map.Add(nums[i], i);
    //     }
    // }
    // return new int[]{};

    // var map = new Dictionary<int, int>();
    // for (int i = 0; i < nums.Length; i++)
    //     if (map.TryGetValue(target - nums[i], out var j))
    //         return new[] { j, i };
    //     else
    //         map[nums[i]] = i;

    // return Array.Empty<int>();

     Dictionary<int, int> map = new Dictionary<int, int>();

    for(int i = 0; i < nums.Length; i++) {
        int complement = target - nums[i];
        if(map.ContainsKey(complement)) {
            return new int[] { map[complement], i };
        }
        map[nums[i]] = i;
    }
    return new int[] { };
  }
}

// Ejemplo: programa con Main para probar Two Sum
class Program
{
    static void Main(string[] args)
    {
        var solution = new Solution();
        int[] nums = { 2, 7, 11, 15 };
        int target = 9;
        int[] result = solution.TwoSum(nums, target);
        System.Console.WriteLine($"[{result[0]}, {result[1]}]"); // [0, 1]
    }
}

File: Python/1. Two Sum.py
from typing import List
from itertools import combinations

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Brute force (comentado para usar hash map)
        # for i in range(len(nums)):
        #     for j in range(i + 1, len(nums)):
        #         if (nums[i] + nums[j] == target):
        #             return [i, j]

    # map = {}
    # for i in range(len(nums)):
    #     complement = target - numsp[i]
    #     if complement in map:
    #         return [map[complement], i]
    #     map[nums[i]] = i    

    # numMap = {}
    # for i, num in enumerate(nums):
    #     numMap[num] = i
    # for i, num in enumerate(nums):
    #     complement = target - num
    #     if complement in numMap and numMap[complement] != i:
    #         return [i, numMap[complement]]


    # pos = {x: i for i, x in enumerate(nums)}

    #     for i, x in enumerate(nums):
    #         need = target - x
    #         if need in pos and pos[need] != i:
    #             return [i, pos[need]]

    # for (i, a), (j, b) in combinations(enumerate(nums), 2):
    #         if a + b == target:
    #             return [i, j]

    # arr = [(v, i) for i, v in enumerate(nums)]
    #     arr.sort()

    #     l, r = 0, len(arr) - 1
    #     while l < r:
    #         s = arr[l][0] + arr[r][0]
    #         if s == target:
    #             return [arr[l][1], arr[r][1]]
    #         if s < target:
    #             l += 1
    #         else:
    #             r -= 1




        numMap = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in numMap:
                return [numMap[complement], i]
            numMap[num] = i
        return []

if __name__ == "__main__":
    solution = Solution()
    nums = [2, 7, 11, 15]
    target = 9
    result = solution.twoSum(nums, target)
    print(result)  # [0, 1]